File: go/array/229_majority_ele.go
package array

func majorityElement_1(nums []int) []int {
	// hash map 忽略 不写

	// 摩尔投票法
	// 1. 第一遍遍历选出可能是超过k/n(本题中1/3) 的元素
	// 2. 第二遍遍历计数确实是否大于1/3
	var element1, vote1 int
	var element2, vote2 int

	for _, v := range nums {

		if vote1 > 0 && v == element1 { // 第一个元素
			vote1++
		} else if vote2 > 0 && v == element2 { // 第二个元素
			vote2++
		} else if vote1 == 0 {
			element1 = v
			vote1++
		} else if vote2 == 0 {
			element2 = v
			vote2++
		} else {
			vote1--
			vote2--
		}
	}

	var cnt1, cnt2 int
	// 第二遍遍历确定 投票出来的元素是超过1/3个还是随机下来的元素
	for _, v := range nums {
		if vote1 > 0 && v == element1 {
			cnt1++
		}
		if vote2 > 0 && v == element2 {
			cnt2++
		}
	}

	result := make([]int, 0)
	if vote1 > 0 && cnt1 > len(nums)/3 {
		result = append(result, element1)
	}

	if vote2 > 0 && cnt2 > len(nums)/3 {
		result = append(result, element2)
	}
	return result
}

File: rust/src/array/229_majority_ele.rs
pub fn majority_element_1(nums: Vec<i32>) -> Vec<i32> {

    let (mut element1, mut vote1) = (0, 0);
    let (mut element2, mut vote2) = (0, 0);

    for &v in nums.iter() {
        if vote1 > 0 && v == element1 {
            vote1 += 1;
        }else if vote2 > 0 && v == element2 {
            vote2 += 1;
        }else if vote1 == 0 {
            element1 = v;
            vote1 += 1;
        }else if vote2 == 0 {
            element2 = v;
            vote2 += 1;
        }else {
            vote1 -= 1;
            vote2 -= 1;
        }
    }

    let (mut cnt1, mut cnt2) = (0, 0);

    for &v in nums.iter() {
        if vote1 > 0 && v == element1 {
            cnt1 += 1;
        }
        if vote2 > 0 && v == element2 {
            cnt2 +=1;
        }
    }

    let mut result = vec![];

    if cnt1 > nums.len() / 3 {
        result.push(element1);
    }
    if cnt2 > nums.len() / 3 {
        result.push(element2);
    }

    result
}

pub fn majority_element_2(nums: Vec<i32>) -> Vec<i32> {

    // rusty 的写法

    let (mut element1, mut vote1) = (0, 0);
    let (mut element2, mut vote2) = (0, 0);


    for &v in nums.iter() {
        match (vote1 > 0, v == element1, vote2 > 0, v == element2) {
            (true, true, _, _) => vote1 +=1,
            (_, _, true, true) => vote2 +=1,
            (false, _, _, _) => {element1 = v; vote1 += 1},
            (_,_,false, _)  => {element2 = v; vote2 += 1},
            _ => {vote1 -= 1; vote2 -= 1}
        }
    }

    // let mut ans = Vec::new();
    let mut ans = vec![];

    if vote1 > 0 && nums.iter().filter(|&x| *x == element1).count() > nums.len() / 3 {
        ans.push(element1);
    }

    if vote2 > 0 && nums.iter().filter(|&&x| x == element2).count() > nums.len() / 3 {
        ans.push(element2);
    }

    ans
   
}